SerialPrograms/Source/PokemonSwSh/Programs/DateSpamFarmers/PokemonSwSh_DateSpam-BerryFarmer2.cpp

```diff
@@ -0,0 +1,275 @@
+        //  Warning: This callback will be run from a different thread than this function.
+        return true;
+        });
+
+    StandardBattleMenuWatcher battle_menu_detector(false);
+    StartBattleWatcher start_battle_detector;
+
+    Rustling result = Rustling::No;
+    int ret = run_until(
+        env.console, context,
+        [&](BotBaseContext& context) {
+            pbf_wait(context, RUSTLING_TIMEOUT);
+            context.wait_for_all_requests();
+        },
+        { {initial_rustling_detector}, {battle_menu_detector}, {start_battle_detector} }
+        );
+
+    if (ret == 0) {
+        env.console.log("BerryFarmer: Initial Rustling detected.");
+        initial_rustling_time = std::chrono::system_clock::now();
+
+        result = Rustling::Slow;
+        
+        int ret1 = run_until(
+            env.console, context,
+            [&](BotBaseContext& context) {
+                pbf_wait(context, RUSTLING_TIMEOUT);
+                context.wait_for_all_requests();
+            },
+            { {secondary_rustling_detector} }
+            );
+
+        if (ret1 == 0) {
+            env.console.log("BerryFarmer: Secondary Rustling detected.");
+            std::chrono::time_point secondary_rustling_time = std::chrono::system_clock::now();
+            if (std::chrono::duration_cast<std::chrono::milliseconds>(secondary_rustling_time - initial_rustling_time).count() <= RUSTLING_INTERVAL) {
+                result = Rustling::Fast;
+            }
+        }
+
+    }
+    else if (ret == 1) {
+        env.console.log("BerryFarmer: Battle Menu detected!");
+        result = Rustling::Battle;
+    } 
+    else if (ret == 2) {
+        env.console.log("BerryFarmer: Battle Start detected.");
+        // waiting here so we can be sure the battle is fully underway so escaping is possible
+        pbf_wait(context, START_BATTLE_DELAY);
+        context.wait_for_all_requests();
+        result = Rustling::Battle;
+    }
+    else {
+        result = Rustling::No;
+    }
+        
+    context.wait_for_all_requests();
+    return result;
+}
+
+uint32_t BerryFarmer2::do_secondary_attempts(SingleSwitchProgramEnvironment& env, BotBaseContext& context, Rustling rustling) {
+    uint8_t no_rustling = (rustling == Rustling::No) ? 1 : 0;
+    Rustling current_rustling = rustling;
+    uint32_t attempts = 0;
+
+    while ((current_rustling == Rustling::Slow || current_rustling == Rustling::No) && no_rustling < 3) { 
+        /* Slow or No rustling, not in Battle! */
+        pbf_mash_button(context, BUTTON_ZL, 240);
+        pbf_mash_button(context, BUTTON_B, 10);
+        attempts++;
+
+        current_rustling = check_rustling(env, context);
+
+        if (current_rustling == Rustling::No) {
+            no_rustling++;
+        }
+    }
+    /* Fast rustling, in Battle or too many times No rustling */
+    if (no_rustling >= 3) {
+        return attempts;
+    }
+    if (current_rustling == Rustling::Fast) {
```

Comment:
If I'm reading this right, you're shaking the tree even though it's fast rustling. (thus risking a battle) Is this somehow a calculated optimal algorithm?

Reply:
Yes, if there is at least one slow rustling first there will be an additional fetch attempt at the end. While I don't have exact numbers it is quite rare for a battle to happen after the first fast rustling (unless it's right after the very first fetch attempt).
This way should be slightly more efficient if battles are less than one in 20ish extra attempts. 
During my tests battles were less than one in 40.